sdl_viewer/src/terrain_drawer/layer.rs

```diff
@@ -0,0 +1,233 @@
+    pub fn new<P: AsRef<std::path::Path>>(
+        program: &GlProgram,
+        path: P,
+        texture_size: u32,
+    ) -> io::Result<Self> {
+        assert!(texture_size % 2 == 0 && texture_size > 0);
+        let metadata = Metadata::from_dir(&path)?;
+        let (height_tiles, color_tiles) = metadata.read_tiles(&path)?;
+
+        let grid_coordinates = GridCoordinateFrame::new(program, metadata, texture_size);
+
+        // Initial terrain pos
+        let terrain_pos: Vector2<i64> =
+            grid_coordinates.terrain_pos_for_camera_pos(Vector3::zero());
+        let u_terrain_pos = GlUniform::new(&program, "terrain_pos", terrain_pos.cast().unwrap());
+
+        let height_initial = height_tiles.load(
+            terrain_pos.x,
+            terrain_pos.y,
+            texture_size.try_into().unwrap(),
+            texture_size.try_into().unwrap(),
+        );
+        let heightmap = GlMovingWindowTexture::new(
+            &program,
+            Rc::clone(&program.gl),
+            "height",
+            texture_size,
+            0, // texture_unit
+            height_initial,
+        );
+
+        let color_initial = color_tiles.load(
+            terrain_pos.x,
+            terrain_pos.y,
+            texture_size.try_into().unwrap(),
+            texture_size.try_into().unwrap(),
+        );
+        let colormap = GlMovingWindowTexture::new(
+            &program,
+            Rc::clone(&program.gl),
+            "color",
+            texture_size,
+            1, // texture_unit
+            color_initial,
+        );
+        let texture_size = i64::from(texture_size);
+
+        Ok(TerrainLayer {
+            grid_coordinates,
+            terrain_pos,
+            u_terrain_pos,
+            height_tiles,
+            color_tiles,
+            heightmap,
+            colormap,
+            texture_size,
+        })
+    }
+
+    // We already have the data between self.terrain_pos and self.terrain_pos + texture_size
+    // Only fetch the "L" shape that is needed, as separate horizontal and vertical strips.
+    // Don't get confused, the horizontal strip is determined by the movement in y direction and
+    // the vertical strip is determined by the movement in x direction.
+    pub fn update(&mut self, cur_world_pos: Vector3<f64>) {
+        let cur_pos = self
+            .grid_coordinates
+            .terrain_pos_for_camera_pos(cur_world_pos);
+        let moved = cur_pos - self.terrain_pos;
+        self.terrain_pos = cur_pos;
+        // Convert to f64, because GLSL doesn't understand i64
+        assert!(
+            cur_pos.x < F64_MAX_SAFE_INT && cur_pos.x > F64_MIN_SAFE_INT,
+            "Terrain location not representable."
+        );
+        assert!(
+            cur_pos.y < F64_MAX_SAFE_INT && cur_pos.y > F64_MIN_SAFE_INT,
+            "Terrain location not representable."
+        );
+        self.u_terrain_pos.value = cur_pos.cast().unwrap();
+
+        let hori_strip = if moved.y > 0 {
```

Comment:
How about something like:
```
let (pos_y, size_y) = if moved.y > 0 {
     (self.terrain_pos.y + self.texture_size, moved.y.try_into().unwrap())
} else {
    (cur_pos.y, moved.y.abs().try_into().unwrap())
};
let hori_strip = self.load(cur_pos.x, pos_y, self.texture_size.try_into().unwrap(), size_y);
```
? Similarly below.

Reply:
Tbh I find the current version a little easier on the brain, just because there's fewer local variables. Do you find it clearer or is it about not repeating code?

Reply:
It's just a little more concise. If you find the current version easier to read, let's keep it.